Quick and dirty matmul

charris opened the pull request:
Adds `__matmul__` and `__rmatmul__`. This is a work in progress and currently lacks `__imatmul__` and tests. The methods seem to work, but python3.5 fails to call them for the `@` operator. I suspect this may be related to ndarray being a class defined in C, as a subclass with the operators added does seem to work.

The methods use einsum for the stacked case and types not available in cblas, consequently they do not work for object arrays. However, they are faster for integer arrays of non-trivial size than the current dot.

njsmith commented:
Yeah, for a class defined in C you have to fill in the new `nb_matrix_multiply` and `nb_inplace_matrix_multiply` methods that were added to the end of the `PyNumberMethods` struct:
    https://docs.python.org/3.5/c-api/typeobj.html#c.PyNumberMethods

charris commented:
@njsmith Thought so, but I couldn't find it in a quick grep through the Python source. Now I need to worry about left and right and NotImplented. Ugh.

charris commented:
And it means it need to be tested with Python3.5.

njsmith commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -2369,6 +2471,50 @@ array_vdot(PyObject *NPY_UNUSED(dummy), PyObject *args)
+    }
+
+    errval = PyUFunc_CheckOverride(cached_npy_matmul, "__call__",
+                                   args, kwds, &override, 2);
```
I feel like `PyArray_Matmul` should also do the override check, i.e. it should have identical semantics to the `matmul` function?

charris replied:
That raises the question if it should be in the C-API at this point. I'm tempted to leave it out and just go with the `matmul` method.

njsmith replied:
I was also feeling a little hesitant about adding it to the C api, so +1 on
waiting until we're more sure what we're doing here.
On May 14, 2015 11:32 PM, "Charles Harris" <notifications@github.com> wrote:

> In numpy/core/src/multiarray/multiarraymodule.c
> <https://github.com/numpy/numpy/pull/5878#discussion_r30387829>:
>
> > +    PyObject *v, *a, *o = NULL;
> > +    char* kwlist[] = {"a", "b", "out", NULL };
> > +
> > +    if (cached_npy_matmul == NULL) {
> > +        PyObject *module, *dict, *matmul;
> > +
> > +        module = PyImport_ImportModule("numpy.core.multiarray");
> > +        dict = PyModule_GetDict(module);
> > +        matmul = PyDict_GetItemString(dict, "matmul");
> > +        cached_npy_matmul = (PyUFuncObject*)matmul;
> > +        Py_INCREF(cached_npy_matmul);
> > +        Py_DECREF(module);
> > +    }
> > +
> > +    errval = PyUFunc_CheckOverride(cached_npy_matmul, "__call__",
> > +                                   args, kwds, &override, 2);
>
> That raises the question if it should be in the C-API at this point. I'm
> tempted to leave it out and just go with the matmul method.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5878/files#r30387829>.
>

charris replied:
Removed.

charris commented:
Works in Python3.5 now.
```
>>> import numpy as np
>>> x = np.ones(3)
>>> m = np.eye(3)
>>> x @ x
3.0
>>> x @ [1,2,3]
6.0
>>> [1,2,3] @ x
6.0
>>> m @ x
array([ 1.,  1.,  1.])
>>> x @ m
array([ 1.,  1.,  1.])
>>> x @ 1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Scalar operands are not allowed, use '*' instead
```

njsmith commented:
Woohoo!

Should probably be a ValueError I guess?
On May 14, 2015 11:32 PM, "Charles Harris" <notifications@github.com> wrote:

> Works in Python3.5 now.
>
> >>> import numpy as np
> >>> x = np.ones(3)
> >>> m = np.eye(3)
> >>> x @ x
> 3.0
> >>> x @ [1,2,3]
> 6.0
> >>> [1,2,3] @ x
> 6.0
> >>> m @ x
> array([ 1.,  1.,  1.])
> >>> x @ m
> array([ 1.,  1.,  1.])
> >>> x @ 1
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> TypeError: Scalar operands are not allowed, use '*' instead
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5878#issuecomment-102283129>.
>

charris commented:
ValueError used.

charris commented:
Seems to work with stacks as it should.
```
>>> mm = np.stack([m]*2, axis=0)
>>> mm @ mm
array([[[ 1.,  0.,  0.],
        [ 0.,  1.,  0.],
        [ 0.,  0.,  1.]],

       [[ 1.,  0.,  0.],
        [ 0.,  1.,  0.],
        [ 0.,  0.,  1.]]])
```

pv commented:
Can you remind why this cannot be implemented as a gufunc, which would get `__numpy_ufunc__` and the outer iterators for free? I guess this has to do with 1d vector issues?

charris commented:
@pv It can be, I have an old branch that makes a start on that, which is why this is quick and dirty ;)

charris committed:
TST: Add tests for multiarray.matmul and operator '@'.

njsmith commented:
I don't object to shipping something quick-and-dirty in 1.10, but I'll note
in case anyone is interested that from a quick look it doesn't seem like
it'd be that hard to enhance gufuncs to support "i?j,jk?->i?k?" syntax I
suggested in some mailing list message somewhere (where "?" means "this
axis is optional -- if not present in the input, treat it as 1, and then
remove it from the output". The gufunc code actually has some checks we
just added since 1.9 that have no purpose except to *prevent* it from
treating missing axes as if they had size 1, so making this work would
mostly be a matter of bookkeeping -- adding the ? syntax to the signature
parser, making a note somewhere about which dimensions were marked as
optional, etc.

On Fri, May 15, 2015 at 7:01 AM, Charles Harris <notifications@github.com>
wrote:

> @pv <https://github.com/pv> It can be, I have an old branch that makes a
> start on that, which is why this is quick and dirty ;)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5878#issuecomment-102406776>.
>



-- 
Nathaniel J. Smith -- http://vorpus.org

jaimefrio commented:
To keep track of the optional dimensions we need to add more items to the `PyUfuncObject` struct. And to do that without breaking ABI compatibility we need to hide away its contents, I guess with something similar to the `PyArrayObject_fields` one. At some point in time @juliantaylor mentioned he was working on the hiding, and I put that work in the backburner until then.

njsmith commented:
While hiding it away is doubtless a good idea, we're already adding new items to `PyArray_Descr` in 1.10 (see #5343), so I don't know that it matters much if we add new items to another struct as well. Plus even in the worst case this only breaks forward compatibility, not backward, and [Ralf says](https://github.com/numpy/numpy/issues/5888#issuecomment-102848230) that we outright enforce lack-of-forward-compatibility regardless.

charris commented:
I actually wrote four gufuncs for this last year, together with some changes to the ufunc generator to make them easier to declare. Ideally BLAS needs to always be available, so we really need the `blas_lite` fallback position and for that I would want to move the code from the linalg package to somewhere common. I also don't like the mutual dependence of the multiarray and ufunc modules, but can live with it.

This PR provides tests and a chance for people to fool with the operator and provide feedback, I don't think Python 3.5 will be standard for another year or so. One problem I've run into is that the operator overrides aren't working as they should, although the matmul function does just fine. I was looking into that when a big chunk of my cpu memory crapped out, so until I put in new memory and get hold the work sitting on disk I'm a bit crippled, probably until Wednesday or so if I'm lucky.

njsmith commented:
The problem is that it's pretty fragile to try patching this together out
of multiple gufuncs for the different shapes, because there is stuff we are
supposed to do on the way into the gufunc (e.g., __numpy_ufunc__
delegation) *before* we can safely cast the inputs to become arrays, and
thus before we know what shape the array is and which gufunc to dispatch
to. There is some extremely awkward code in np.linalg.solve to try and work
around this already. Teaching the gufunc code how to handle this directly
would simplify everything a lot.
.
Concur on the mutual dependence between umath and multiarray being nasty.
In the long run I still think we should merge them into one library, but in
the mean time... Is there anything in multiarray that actually needs blas,
though, besides dot? We could just move the blas dependency entirely to
umath, no? Instead of having dot be handled through the dtype ArrFuncs, we
could just use regular ufunc loop dispatch? And then we'd still need a way
for multiarray to get a handle to the dot gufunc, but that doesn't require
any *new* kluges, it can use the same hack that multiarray uses to get a
handle to the add ufunc for +.
.
Not sure what you mean about operator overrides, but note that how they
*ought* to work is under heavy discussion right now in #5844 so depending
on what the issue is you might not want to worry about it until that's
resolved to worry about it...

On Mon, May 18, 2015 at 12:59 AM, Charles Harris <notifications@github.com>
wrote:

> I actually wrote four gufuncs for this last year, together with some
> changes to the ufunc generator to make them easier to declare. Ideally BLAS
> needs to always be available, so we really need the blas_lite fallback
> position and for that I would want to move the code from the linalg package
> to somewhere common. I also don't like the mutual dependence of the
> multiarray and ufunc modules, but can live with it.
>
> This PR provides tests and a chance for people to fool with the operator
> and provide feedback, I don't think Python 3.5 will be standard for another
> year or so. One problem I've run into is that the operator overrides aren't
> working as they should, although the matmul function does just fine. I was
> looking into that when a big chunk of my cpu memory crapped out, so until I
> put in new memory and get hold the work sitting on disk I'm a bit crippled,
> probably until Wednesday or so if I'm lucky.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5878#issuecomment-102955438>.
>



-- 
Nathaniel J. Smith -- http://vorpus.org

charris committed:
ENH: Add a matmul function to multiarray

This is the functional counterpart of the '@' operator that will be
available in Python 3.5 with the addition of an out keyword. It
operates like the dot function except that

- scalar multiplication is not allowed.
- multiplication of arrays with more than 2 dimensions broadcasts.

The last means that when arrays have more than 2 dimensions they are
treated as stacks of matrices and those stacks are broadcast against
each other unlike the current behavior of dot that does an outer
product. Like dot, matmul is aware of `__numpy_ufunc__` and can be
overridden.

The current version of the function uses einsum when cblas cannot be
used, hence object arrays are not yet supported.

charris committed:
ENH: Add support for the '@' operator added in Python 3.5.

charris committed:
STY: Minor C style fixes for multiarraymodule.c.

Start labels at beginning of line.
Break a long line
Whitespace.

charris committed:
DOC: Document '@' and matmul.

Document the matmul function and add '@' to the operator
section of the reference manual.

charris committed:
DOC: Mention support for the '@' operator in the 1.10 release notes.

charris commented:
I think this is ready to go pending tests. The override mechanism is the same as currently implemented for other functions, we can fix it up when the issue is decided and a transition path laid out.

charris commented:
Take that back, there are still some problems with the operator.

charris commented:
Should all be fixed now.

charris commented:
I'm going to merge this in a day or two if noone beets me to it and there are no strenuous objections.

charris committed:
BUG: Fix npy_cache_pyfunc to properly implement inline.

The function was not static, which led to multiple definition
errors.
